### PR TITLE
bpo-43984: Update winreg.SetValueEx to make sure the value set is not -1

### PR DESCRIPTION
**Pull request by shreyanavigyan**
[bpo-43984](https://bugs.python.org/issue43984): Update winreg.SetValueEx to make sure the value set is not -1

- [x] Fix `REG_QWORD` case in `Py2Reg`
- [x] Fix `REG_DWORD` case in `Py2Reg`
- [x] Add test
- [ ] Make sure the test does not appear as virus to the antivirus software!

https://bugs.python.org/issue43984

### CODE REVIEW

**Comment by shreyanavigyan** [comment]
Skip news

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c:582]: I don't know this part of the code base, _but_ it does look like this leaks the memory allocated on line 570. Ditto for the change at line 603.

Comment (shreyanavigyan) [reply PC/winreg.c:582]: It doesn't leak memory. The problem is that if an error occurred in PyLong_AsUnsignedLong (like passing a negative python integer), PyLong_AsUnsignedLong returns -1 and the code here think -1 is the value to be set and sets it. This PR solves that problem.

Comment (erlend-aasland) [reply PC/winreg.c:582]: Looks like a leak to me; the caller does not free the memory if `FALSE` is returned :)

Comment (erlend-aasland) [reply PC/winreg.c:582]: cc. @eryksun

Comment (pfmoore) [reply PC/winreg.c:582]: Agreed - `retDataBuf` isn't getting freed in that case.

Comment (shreyanavigyan) [reply PC/winreg.c:583]: Ooops. Didn't notice that. Just freeing the memory using `free` will do, right?

Comment (shreyanavigyan) [reply PC/winreg.c:582]: I'm fixing that. Wait.

Comment (erlend-aasland) [reply PC/winreg.c:582]: @pfmoore Also, there seems to be inconsistent behaviour for the various cases; for `REG_EXPAND_SZ` memory is only allocated if `value == Py_None`. Don't know if it matters, though :)

Comment (erlend-aasland) [reply PC/winreg.c:582]: @shreyanavigyan Either free the memory explicitly before returning, or just move the conversion up before the allocation. I'd prefer the latter; for my own code, I prefer deferring allocation as late as possible in order to make it easier to clean up. But that's just personal preference :)

Comment (erlend-aasland) [reply PC/winreg.c:582]: Also, this part of the code base seems to use the deprecated `PyMem_FREE`/`PyMem_NEW` iso. `PyMem_Free`/`PyMem_New`.

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]:
```suggestion
                    PyMem_Free(retDataBuf);
```

Comment (erlend-aasland) [reply PC/winreg.c]: Ditto for the other `free`.

Comment (shreyanavigyan) [reply PC/winreg.c]: Fixed in commit 78417143294ff843d58dd61d1fb7e190227566b5

**Comment by shreyanavigyan** [comment]
Can anyone please add a skip news label to this PR?

File: PC/winreg.c
Comment (erlend-aasland) [reply PC/winreg.c]: Thanks :)

File: PC/winreg.c
Comment (shreyanavigyan) [reply PC/winreg.c:582]: > Also, this part of the code base seems to use the deprecated `PyMem_FREE`/`PyMem_NEW` iso. `PyMem_Free`/`PyMem_New`.

I'll open a different issue and PR for that if needed.

Comment (erlend-aasland) [reply PC/winreg.c:582]: See GH-23586 for a previous cleanup, @shreyanavigyan.

**Comment by erlend-aasland** [comment]
> Can anyone please add a skip news label to this PR?

Isn't this a change of behaviour? If so, I'd say it calls for a NEWS item.

**Comment by shreyanavigyan** [comment]
News entry added

**Comment by erlend-aasland** [comment]
Perhaps you could use the bug report to write a test case? (Lib/test/test_winreg.py)

File: PC/winreg.c
Comment (eryksun) [reply PC/winreg.c:582]: > move the conversion up before the allocation. 

Seems like a good idea to me. For example:

```C
        case REG_DWORD: {
            DWORD d = 0;
            if (PyLong_Check(value)) {
                d = PyLong_AsUnsignedLong(value);
                if (d == (DWORD)(-1) && PyErr_Occurred()) {
                    return FALSE;
                }
            } else if (value != Py_None) {
                return FALSE;
            }
            *retDataBuf = (BYTE *)PyMem_NEW(DWORD, 1);
            if (*retDataBuf == NULL) {
                PyErr_NoMemory();
                return FALSE;
            }
            memcpy(*retDataBuf, &d, sizeof(DWORD));
            *retDataSize = sizeof(DWORD);
            break;
        }
        case REG_QWORD: {
            DWORD64 d = 0;
            if (PyLong_Check(value)) {
                d = PyLong_AsUnsignedLongLong(value);
                if (d == (DWORD64)(-1) && PyErr_Occurred()) {
                    return FALSE;
                }
            } else if (value != Py_None) {
                return FALSE;
            }
            *retDataBuf = (BYTE *)PyMem_NEW(DWORD64, 1);
            if (*retDataBuf == NULL) {
                PyErr_NoMemory();
                return FALSE;
            }
            memcpy(*retDataBuf, &d, sizeof(DWORD64));
            *retDataSize = sizeof(DWORD64);
            break;
        }
```

Comment (shreyanavigyan) [reply PC/winreg.c:582]: Should I implement this? (I've already added a commit to free up retDataBuf before returning FALSE. Is it worth changing the whole code control flow?)

Comment (erlend-aasland) [reply PC/winreg.c:582]: IMO it's a better solution (avoids unneeded malloc/free). I leave it for Eryk and Paul to decide :)

Comment (pfmoore) [reply PC/winreg.c:582]: I agree. There's no requirement to touch as few lines of code as possible with a change, so let's use the cleanest approach we can to get the required behaviour.

Comment (shreyanavigyan) [reply PC/winreg.c:582]: Should I proceed to change the code?

Comment (shreyanavigyan) [reply PC/winreg.c:582]: Looking at the existing code again. It seems changing it would be a better approach. I'm applying the changes that @eryksun suggested.

**Comment by mehdi40277** [comment]
at the time of the initial coding, I defined using a formula that had numerical coordinates and each currency code had a unique identity, and at the same time was operating cash under the pseudonym Satoshi Nakamoto. And my goal The induction of banking stability cryptocurrency was created to elevate a revolution in the attitude of the central treasury and the monetary organization, and the idea of ​​creating each sell was in line with a coding program that would create economic transparency in the world and provide financial security to capital owners. Every human being has a new vision and a true belief that the honesty of the thoughts that originate leads the heart to the reasoning of the metaphysical science of the brain. During these several years since the birth of this financial system, many people have abused my position and name and shown themselves in my place. Wow, the growth attitude is to induce the removal of Satoshi Nakamoto and name an identity that, if his name produced an economic revolution, stemmed only from faith and God's direct connection to the intrinsic intellect and intuition that served human society. But I can no longer bear that Such capital that using the sciences of God This kind of plunder is created by opportunists who try to shake this illusion from the darkest parts of everything. The goal is to create the attitude of the birth of superior minds to leave a legacy to the future, as our past left us. In fact, man The further it goes, the closer it actually gets to the past of its predecessors, and this means that we will exist again and again in nature, but with a different appearance and mask that we choose. So the new revolution is the attitude of Bitcoin today towards It starts a new one, and that is creating the business balance of an economic act in relation to certain times, such as the coronary heart disease epidemic in the world, which with this excellence creates the confidence of capital owners in times of such problems and disasters and the existence of Bitcoin financial insurance. The title of Quinn Insurance, which is based on stock exchange transactions, of course, depending on the type of problem, for example, during major illness, the activity to provide medicine, vaccines and items needed by human societies, because the growth of a nation is only when that nation is equal to gender and Race to know that humans based on reason and Judy is judged, not visually. So my friends, brothers, sisters and masters. I once again, as a service to you, proves the motivation of time and not language. So my heart is the place of love and wealth that you have in your heart and You give it to your fellow man with a smile. You are your real wealth. Love is believed to be a seedling in the existence of every hero.

File: Misc/NEWS.d/next/Windows/2021-05-02-15-29-33.bpo-43984.U92jiv.rst
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Windows/2021-05-02-15-29-33.bpo-43984.U92jiv.rst]: Perhaps you could phrase this so it's easily understandable for all users. Maybe something like:
```
:meth:`winreg.SetValueEx` now leaves the target register untouched in case of conversion errors. Previously, ``-1`` would be written in case of such errors.
```

Take a look at [section 7 in the dev guide](https://devguide.python.org/documenting/#restructuredtext-primer) for more examples of how to use reST markup :)

Comment (shreyanavigyan) [reply Misc/NEWS.d/next/Windows/2021-05-02-15-29-33.bpo-43984.U92jiv.rst]: Sure. I'm now working on implementing all the changes proposed in this PR.

Comment (erlend-aasland) [reply Misc/NEWS.d/next/Windows/2021-05-02-15-29-33.bpo-43984.U92jiv.rst]: Great! Also remember to add a test target this branch.

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]: Please follow PEP 7, ref. section 3.5 in the dev guide:

```suggestion
            }
            else if (value != Py_None) {
```

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]: Please follow PEP 7, ref. section 3.5 in the dev guide:

```suggestion
            }
            else if (value != Py_None) {
```

File: Misc/NEWS.d/next/Windows/2021-05-02-15-29-33.bpo-43984.U92jiv.rst
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Windows/2021-05-02-15-29-33.bpo-43984.U92jiv.rst]: It should be noted that this change only affects `REG_DWORD` and `REG_QWORD` values.

**Review by erlend-aasland — CHANGES_REQUESTED** [review]

File: PC/winreg.c
Comment (erlend-aasland) [reply PC/winreg.c]: ~Also, it seems like the indent is off. Please check that it is a multiple of four (spaces only, no tabs).~

File: Misc/NEWS.d/next/Windows/2021-05-02-15-29-33.bpo-43984.U92jiv.rst
Comment (shreyanavigyan) [reply Misc/NEWS.d/next/Windows/2021-05-02-15-29-33.bpo-43984.U92jiv.rst]: No problem. But let me add the tests first.

Comment (tjguk) [reply Misc/NEWS.d/next/Windows/2021-05-02-15-29-33.bpo-43984.U92jiv.rst]: As a point of English: "in case of" here doesn't mean what you want it to. "in case of ..." means: it leaves it untouched in every situation, and the reason is that conversion errors might happen. What you want is "in the case of ..." which means: if a conversion error happens, the value is left untouched.

Comment (erlend-aasland) [reply Misc/NEWS.d/next/Windows/2021-05-02-15-29-33.bpo-43984.U92jiv.rst]: Thanks for pointing that out, @tjguk! It was my suggestion, but I'm no native English speaker :)

File: Lib/test/test_winreg.py
Comment (shreyanavigyan) [inline Lib/test/test_winreg.py]: There's something wrong with this test I added. My antivirus is identifying this piece of code as virus. Any thoughts?

Comment (shreyanavigyan) [reply Lib/test/test_winreg.py]: If any maintainers figure out the problem please add the required commit. Who cannot add commits please suggest the patch, I'll commit it.

File: Misc/NEWS.d/next/Windows/2021-05-02-15-29-33.bpo-43984.U92jiv.rst
Comment (eryksun) [inline Misc/NEWS.d/next/Windows/2021-05-02-15-29-33.bpo-43984.U92jiv.rst]: You mean "target value", not "target register". In computing, a "register" is a unit of memory in a processor.

File: Lib/test/test_winreg.py
Comment (eryksun) [inline Lib/test/test_winreg.py]:
```suggestion
                with self.assertRaises(FileNotFoundError):
                    QueryValueEx(ck, "test_name")
```
The return value of `QueryValueEx` is a tuple of the form `(value, type_id)`, not a key, and it's not a context manager.

File: Misc/NEWS.d/next/Windows/2021-05-02-15-29-33.bpo-43984.U92jiv.rst
Comment (shreyanavigyan) [reply Misc/NEWS.d/next/Windows/2021-05-02-15-29-33.bpo-43984.U92jiv.rst]: Oops. I'll fix that tomorrow.

**Comment by shreyanavigyan** [comment]
I'm changing the control flow a little bit. It seems we forgot about the case where value can Py_None.

**Comment by shreyanavigyan** [comment]
> I didn't forget the case of `None` in the example code that I gave you. The value of `d` was initialized to 0 for that case.

Oh. But anyway the updated control flow is simple. I don't think we need to change it. But can you help me figure out what's wrong with this test or code? My antivirus is marking it as virus. My antivirus never marks anything virus unless it's very suspicious.

**Comment by eryksun** [comment]
> Oh. But anyway the updated control flow is simple. I don't think we need to change it. 

It's your PR. But to explain my choices, I didn't like the redundant check for `Py_None`, and I preferred the common case of the `PyLong_Check` and conversion handled upfront. That's just a reflection of my thought process.

> But can you help me figure out what's wrong with this test or code? My antivirus is marking it as virus. 
> My antivirus never marks anything virus unless it's very suspicious.

Sorry, I can't help with that.

**Comment by shreyanavigyan** [comment]
> Sorry, I can't help with that.
 
It's ok. But I'm not sure my antivirus is going crazy or there's a bug.

File: Lib/test/test_winreg.py
Comment (erlend-aasland) [inline Lib/test/test_winreg.py]: 1. You should assert that the problem you are addressing in the PR is fixed: did `SetValueEx` overwrite the current registry value or not? Currently, you do not assert that the queried value is unchanged.
2. There's no test case for `REG_QWORD`.
3. Nit: I'd shorten the docstring to a one-liner, something a la: _"bpo-43984: assert that registry values are not set to -1 in the case of overflow errors."_

Comment (shreyanavigyan) [reply Lib/test/test_winreg.py]: 1. Queried value wouldn't exist because SetValueEx didn't create or set the key. The assert case here is QueryValueEx would raise a FileNotFoundError because it couldn't find the key.
2. Sure. Will do that in a few minutes.
3. I've followed the same commenting style used in the test cases in test_winreg.

Comment (erlend-aasland) [reply Lib/test/test_winreg.py]: > 1. Queried value wouldn't exist because SetValueEx didn't create or set the key. The assert case here is QueryValueEx would raise a FileNotFoundError because it couldn't find the key.

Makes sense. Sorry 'bout the noise.

Comment (eryksun) [reply Lib/test/test_winreg.py]: > Queried value wouldn't exist because SetValueEx didn't create or set the key. The assert case here is QueryValueEx would raise a FileNotFoundError because it couldn't find the key.

`SetValueEx` and `QueryValueEx` are setting and querying a value, not a key. 

The analog to a filesystem is that a key is like a directory; a value is like a [named stream](https://docs.microsoft.com/en-us/windows/win32/fileio/file-streams) in a directory (e.g. "dirname:streamname:$DATA"); and the value type (e.g. `REG_SZ`) is like a stream type (e.g. $DATA). A directory can be mounted by another filesystem, or a directory can be a symlink (i.e. contain a symlink $REPARSE_POINT stream) that targets another directory. Similarly a key can be mounted by a hive (e.g. "\\Registry\\Machine\\System" is mounted by the hive "%SystemRoot%\\System32\\config\\SYSTEM"), or a key can contain a symlink (`REG_LINK` value type) to another key (e.g. "\\Registry\\User\\\<sid\>\\Software\\Classes" is a symlink to "\\Registry\\User\\\<sid\>_Classes"). Values in a key are not individually secured by a security descriptor, just like named streams in a directory are not individually secured. A significant difference is that values are not accessed as objects that can be the target of an open, unlike named streams in a directory. Thus a value cannot be referenced by a handle, and a value cannot be the target of a symlink.

Comment (shreyanavigyan) [reply Lib/test/test_winreg.py]: Ooops. I always mess up between SetValueEx, QueryValueEx  and SetValue, QueryValue. Anyway the value will never be created and set by SetValueEx. That's what I meant to say.

Comment (erlend-aasland) [reply Lib/test/test_winreg.py]: Thanks for the explanation, Eryk!

Comment (shreyanavigyan) [reply Lib/test/test_winreg.py]: I'm adding the test for REG_QWORD now.

**Comment by shreyanavigyan** [comment]
PR is ready. No more commits will be made.

**Comment by shreyanavigyan** [comment]
A Friendly Ping.

**Comment by zooba** [comment]
Sorry for the lack of review. Unfortunately, it looks like someone added a new required check, so I'll close/reopen to re-run it, but the change looks okay to me. If someone else gets back to this before I do, they can merge.

**Comment by zooba** [comment]
The test failure on Azure Pipelines is patchcheck reporting extra whitespace in the new test. I can't see it in the diff, but I find it's usually trailing whitespace in a comment.

We shouldn't merge with that failure, or it'll show up randomly in the future.

File: Lib/test/test_winreg.py
Comment (AlexWaygood) [inline Lib/test/test_winreg.py]:
```suggestion

```

File: Lib/test/test_winreg.py
Comment (AlexWaygood) [inline Lib/test/test_winreg.py]:
```suggestion

```

File: Lib/test/test_winreg.py
Comment (AlexWaygood) [inline Lib/test/test_winreg.py:342]:
```suggestion

    def test_setvalueex_negative_one_check(self):
```

File: Lib/test/test_winreg.py
Comment (AlexWaygood) [inline Lib/test/test_winreg.py:357]:
```suggestion
            DeleteKey(HKEY_CURRENT_USER, test_key_name)

```

**Comment by zooba** [comment]
Thanks! FWIW, I suspect the antimalware detection was probably because it's an unsigned binary messing with the registry. Once we do our signed release, I'm sure it'll be fine.

**Pull request merged**
